### PR TITLE
Templating Functions & HCL Variables for Artifacts, Deploys

### PR DESCRIPTION
**Pull request by mitchellh**
This introduces two major features (related):

  * A set of a `template`-prefixed HCL functions that template files or directories with the currently available HCL variables. 
  * Artifact, deploy HCL variables for registry, deploy, and release stanzas. You can use this to reference information about the result of the prior operation.

**This PR temporarily disables auth checks completely.** We will add these again at runtime in a future PR, prior to 0.2.0.

Depends on https://github.com/hashicorp/waypoint-plugin-sdk/pull/5 and will update go.mod after merging that.

## Example

```hcl
app "web" {
  build {
    use "docker" {
      dockerfile = templatefile("${path.app}/Dockerfile.prod", {
        custom_var = "hello"
      })
    }

    registry {
      use "docker" {
        image = artifact.image
        tag = gitrefpretty()
      }
    }
  }
  
  # ...
}
```

### CODE REVIEW

File: builtin/docker/copy.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package docker
+
+import (
+	"io"
+	"os"
+)
+
+// From: https://gist.github.com/m4ng0squ4sh/92462b38df26839a3ca324697c8cba04
```
Comment (krantzinator) [inline builtin/docker/copy.go:8]: User moved profiles
``` suggestion
// From: https://gist.github.com/r0l1/92462b38df26839a3ca324697c8cba04
```

**Review by krantzinator — APPROVED** [review]
Exciting :tada:

File: internal/core/eval_context.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	}
+
+	valBytes, ok := val.Interface().([]byte)
+	if !ok {
+		return nil
+	}
+
+	var raw map[string]interface{}
+	if err := json.Unmarshal(valBytes, &raw); err != nil {
+		return err
```
Comment (evanphx) [inline internal/core/eval_context.go:30]: We should probably give them a detailed error here about what parsing failed. Folks are going to pass bare strings accidentally here most likely so we should give them at first. Maybe: `errors.Wrapf(err, "Error parsing TemplateData for '%s'", key)`.

Comment (evanphx) [reply internal/core/eval_context.go:30]: Though maybe I'm mistaken and this would already have been validate before being put into the protobuf?

**Review by evanphx — CHANGES_REQUESTED** [review]
This is a biggy! All looks good, though it's hard to follow some of the changes, like I wasn't sure why all the template functions don't take a basedir anymore? I must have missed something in the reading where that basedir is coming from now. Maybe it's that the base dir is being applied to path elsewhere now?

File: internal/core/eval_context.go (hunk 2 above)
Comment (mitchellh) [reply internal/core/eval_context.go:30]: Yeah, the JSON is guaranteed because we encode it ourselves in the SDK based on the structure. 😄 Still, I can wrap this in the hopefully-impossible case it happens.

**Review by evanphx — APPROVED** [review]
